pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentLoader.java

```diff
@@ -173,6 +174,34 @@ public static ImmutableSegment load(File indexDir, IndexLoadingConfig indexLoadi
     return segment;
   }
 
+  /**
+   * Check segment directory against the table config and schema to see if any preprocessing is needed,
+   * like changing segment format, adding new indices or updating default columns.
+   */
+  public static boolean needPreprocess(SegmentDirectory segmentDirectory, IndexLoadingConfig indexLoadingConfig,
+      @Nullable Schema schema)
+      throws Exception {
+    if (needConvertSegmentFormat(indexLoadingConfig, segmentDirectory.getSegmentMetadata())) {
+      return true;
+    }
+    SegmentPreProcessor preProcessor = new SegmentPreProcessor(segmentDirectory, indexLoadingConfig, schema);
+    return preProcessor.needProcess();
+  }
+
+  @VisibleForTesting
+  static boolean needConvertSegmentFormat(IndexLoadingConfig indexLoadingConfig,
+      SegmentMetadataImpl segmentMetadata) {
+    SegmentVersion segmentVersionToLoad = indexLoadingConfig.getSegmentVersion();
+    if (segmentVersionToLoad == null) {
+      return false;
+    }
+    File indexDir = segmentMetadata.getIndexDir();
+    if (indexDir != null && SegmentDirectoryPaths.segmentDirectoryFor(indexDir, segmentVersionToLoad).isDirectory()) {
+      return false;
+    }
+    return segmentVersionToLoad != segmentMetadata.getVersion();
```

Comment:
Can we make downConversion as NO-OP or throw exception like it does in the actual preprocess code ?

Reply:
Today during segment preprocessor when the converter is created, we throw exception for down convert from v3 to v1 / v2.

 ```
public static SegmentFormatConverter getConverter(SegmentVersion from, SegmentVersion to) {

    if ((from.equals(SegmentVersion.v1) || from.equals(SegmentVersion.v2)) && to.equals(SegmentVersion.v3)) {
      return new SegmentV1V2ToV3FormatConverter();
    }

    throw new UnsupportedOperationException("Unable to find a converter to convert segment from:" + from + " to:" + to);
  }
```

Reply:
For consistent behavior, I agree to continue to throw exception if required to down convert, instead of NOOP, as the upper layer may catch and handle exception.

And I'd prefer to continue to throw the exception from `getConverter`, as per the error msg, it's about not able to find converter to do the down convert, instead of not allowing it.